src/ex2/main.js

```diff
@@ -0,0 +1,146 @@
+        else this.addTodo(value);
+        document.querySelector(".todo__form").reset(); // reset
+      }
+    } catch (error) {
+      console.error(error);
+    }
+  }
+
+  handleDelete(id) {
+    if (confirm(DELETE_CONFIRMATION_MESSAGE)) {
+      this.removeTodoListItem(id);
+      if (!id.includes("error")) this.state.removeTodo(id);
+    }
+  }
+
+  handleClear() {
+    const todoList = this.getTodoList();
+    /**
+     * i check the element count and not the todos array
+     * because i dont save the errors, and if the list
+     * will be full with errors only, the user wont be able to clear
+     */
+    if (todoList.childElementCount > 1)
+      if (confirm(DELETE_CONFIRMATION_MESSAGE)) {
+        todoList.replaceChildren(this.getTodoListItemTemplate());
+        this.state.removeAllTodos();
+      }
+  }
+
+  // pokemon
+  async addPokemonById(id) {
+    if (this.state.isPokemonExist(id)) return alert(POKEMON_ALREADY_EXIST);
+    try {
+      const pokemon = await this.state.addPokemonById(id);
+      this.createAppendTodoListItem(pokemon.getId(), pokemon.getData());
+    } catch (error) {
+      this.addPokemonFail(error);
+    }
+  }
+
+  async addMultiplePokemonsById(ids) {
+    const newIds = this.prepareIds(ids);
+    if (!newIds.length) return alert(ALL_POKEMONS_ALREADY_EXIST);
+    try {
+      const pokemons = await this.state.addMultiplePokemonsById(newIds);
+      pokemons.forEach((pokemon) =>
+        this.createAppendTodoListItem(pokemon.getId(), pokemon.getData())
+      );
+    } catch (error) {
+      this.addPokemonFail(error);
+    }
+  }
+
+  // error -> {message}
+  addPokemonFail({ message }) {
+    if (message.includes("not found") || message.includes("Failed to fetch")) {
+      this.createAppendTodoListItem(`error-${utils.uuidv4()}`, message);
+    } else alert(message);
+  }
+
+  // looking for better name... / split into two functions
+  prepareIds(ids) {
+    /**
+     * split, map to trim
+     * create new set to del dup, spread into array
+     * filter to take out existing pokemons
+     * */
+    const uniqIds = [...new Set(ids.split(",").map((id) => id.trim()))];
+    return uniqIds.filter((id) => !this.state.isPokemonExist(id));
+  }
+
+  // todo
+  addTodo(value) {
+    const todo = new Todo(value);
+    this.createAppendTodoListItem(todo.getId(), todo.getValue());
+    this.state.addTodo(todo);
+  }
+
+  createTodoListItem(id, value) {
+    const item = this.getTodoListItemTemplate().cloneNode(true);
+    item.id = id;
+    item.classList.toggle("hidden");
+    // text - span
+    item.querySelector(".todo__text").innerText = value;
+    // button
+    item
+      .querySelector(".todo__delete")
+      .addEventListener("click", () => this.handleDelete(id));
+    return item;
+  }
+
+  createAppendTodoListItem(id, value, todoList = this.getTodoList()) {
+    todoList.appendChild(this.createTodoListItem(id, value));
+  }
+
+  removeTodoListItem(id, todoList = this.getTodoList()) {
```

Comment:
There's no reason to get this in the function arguments since there's no case where you pass it in, rather you always get it using the function so just define it within the function
```suggestion
  removeTodoListItem(id) {
  const todoList = this.getTodoList();
```

Reply:
fixed.

Reply:
I moved these 3 functions into dom service file

Reply:
fixed.